### PR TITLE
GEM bugfix for issue #20772

### PR DESCRIPTION
**Pull request by jshlee**
@slava77, @kpedro88, @calabria 

The standalone ME0Muon is no longer needed, so i removed it from the standard reco chain.
GEMSegmentBuilder.cc has been fixed.
ME0SegAlgoRU.cc had a warning msg.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
@jshlee while removing the standalone ME0Muon means we won't have a leak in the workflows, if the associated code will persist in CMSSW, I would prefer to have it actually fixed. Someone could reactivate it at any time.

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
resolves #20772

**Comment by kpedro88** [comment]
+1

**Comment by perrotta** [comment]
@jshlee : there is now a conflict in RecoMuon/Configuration/python/RecoMuonPPonly_cff.py, please fix

**Comment by jshlee** [comment]
@perrotta - fixed

**Comment by perrotta** [comment]
please test

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 1]
```diff
@@ -2,75 +2,46 @@
...
+#include "DataFormats/GEMRecHit/interface/ME0SegmentCollection.h"
 
+#include "Geometry/Records/interface/MuonGeometryRecord.h"
+#include "Geometry/CommonTopologies/interface/StripTopology.h"
+#include "Geometry/GEMGeometry/interface/ME0Geometry.h"
+#include "Geometry/GEMGeometry/interface/ME0EtaPartition.h"
+#include "Geometry/GEMGeometry/interface/ME0EtaPartitionSpecs.h"
 
+#include "TrackPropagation/SteppingHelixPropagator/interface/SteppingHelixPropagator.h"
+#include "TrackPropagation/SteppingHelixPropagator/interface/SteppingHelixStateInfo.h"
```
Comment (perrotta) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: Is this include needed?

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 2]
```diff
@@ -2,75 +2,46 @@
...
+#include "TrackingTools/Records/interface/TrackingComponentsRecord.h"
+#include "TrackingTools/AnalyticalJacobians/interface/JacobianCartesianToLocal.h"
+#include "TrackingTools/AnalyticalJacobians/interface/JacobianLocalToCartesian.h"
```
Comment (perrotta) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: Is this include needed?

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 3]
```diff
@@ -2,75 +2,46 @@
...
-#include "Geometry/CommonTopologies/interface/StripTopology.h"
-#include <DataFormats/GeometrySurface/interface/SimpleDiskBounds.h>
-
-
-/** \class ME0SegmentMatcher 
- *
- * \author David Nash
- */
-
+#include "FWCore/PluginManager/interface/ModuleDef.h"
```
Comment (perrotta) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: This include is not needed

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 4]
```diff
@@ -2,75 +2,46 @@
...
+#include "DataFormats/GeometrySurface/interface/LocalError.h"
+#include "DataFormats/MuonReco/interface/ME0Muon.h"
+#include "DataFormats/MuonDetId/interface/ME0DetId.h"
+#include "DataFormats/TrajectoryState/interface/LocalTrajectoryParameters.h"
 #include "DataFormats/TrackReco/interface/TrackFwd.h"
 #include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/GEMRecHit/interface/ME0SegmentCollection.h"
 
+#include "Geometry/Records/interface/MuonGeometryRecord.h"
+#include "Geometry/CommonTopologies/interface/StripTopology.h"
```
Comment (perrotta) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: Is this include needed?

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 5]
```diff
@@ -2,75 +2,46 @@
...
+#include "Geometry/GEMGeometry/interface/ME0Geometry.h"
+#include "Geometry/GEMGeometry/interface/ME0EtaPartition.h"
```
Comment (perrotta) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: Is this include needed?

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 6]
```diff
@@ -2,75 +2,46 @@
...
+#include "Geometry/GEMGeometry/interface/ME0EtaPartitionSpecs.h"
```
Comment (perrotta) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: Is this include needed?

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 7]
```diff
@@ -2,75 +2,46 @@
...
-#include <DataFormats/GEMRecHit/interface/ME0SegmentCollection.h>
-#include <DataFormats/CSCRecHit/interface/CSCSegmentCollection.h>
-
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/Math/interface/AlgebraicROOTObjects.h"
+#include "DataFormats/Math/interface/deltaR.h"
+#include "DataFormats/GeometryVector/interface/GlobalVector.h"
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+#include "DataFormats/GeometrySurface/interface/Plane.h"
+#include "DataFormats/GeometrySurface/interface/SimpleDiskBounds.h"
```
Comment (perrotta) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: Is this include needed?

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 8]
```diff
@@ -2,75 +2,46 @@
...
+#include "DataFormats/GeometrySurface/interface/LocalError.h"
```
Comment (perrotta) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: Is this include needed?

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 9]
```diff
@@ -2,75 +2,46 @@
...
+#include "Geometry/GEMGeometry/interface/ME0EtaPartition.h"
+#include "Geometry/GEMGeometry/interface/ME0EtaPartitionSpecs.h"
 
+#include "TrackPropagation/SteppingHelixPropagator/interface/SteppingHelixPropagator.h"
+#include "TrackPropagation/SteppingHelixPropagator/interface/SteppingHelixStateInfo.h"
+#include "TrackingTools/Records/interface/TrackingComponentsRecord.h"
+#include "TrackingTools/AnalyticalJacobians/interface/JacobianCartesianToLocal.h"
+#include "TrackingTools/AnalyticalJacobians/interface/JacobianLocalToCartesian.h"
 
+#include "TLorentzVector.h"
```
Comment (perrotta) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: Is this include needed?

**Review by perrotta — COMMENTED** [review]
Since you are cleaning this file, you can probably also get rid of the unnneeded includes. I think I have spotted a few: please verify, and remove them if actually not needed

**Comment by perrotta** [comment]
Please modify accordingly also the event content for phase 2:
https://cmssdt.cern.ch/lxr/source/RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py#0086

**Comment by jshlee** [comment]
@perrotta - I've fixed the headers and updated the event content.

**Comment by perrotta** [comment]
please test

**Comment by kpedro88** [comment]
+1

**Comment by perrotta** [comment]
+1
- me0MuonReco sequence Removed from phase 2 reco sequence
- even though not in the standard sequence any more, the memory leaks reported in #20772 are fixed and some cleaning of the code is applied

File: RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc [hunk 10]
```diff
@@ -125,193 +73,161 @@ ME0SegmentMatcher::ME0SegmentMatcher(const edm::ParameterSet& pas) {
...
-      TrajectoryStateOnSurface lastrecostate;
+    p3reco = GlobalVector(thisTrack->outerPx(), thisTrack->outerPy(), thisTrack->outerPz());
+    r3reco = GlobalVector(thisTrack->outerX(), thisTrack->outerY(), thisTrack->outerZ());
 
-      //const SteppingHelixPropagator* ThisshProp = 
-      //dynamic_cast<const SteppingHelixPropagator*>(&*shProp);
+    AlgebraicSymMatrix66 covReco;
+    //This is to fill the cov matrix correctly
+    AlgebraicSymMatrix55 covReco_curv;
+    covReco_curv = thisTrack->outerStateCovariance();
```
Comment (davidlange6) [inline RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: hi @jshlee - can this be a const AlgebraicSymMatrix55 & instead?

Comment (jshlee) [reply RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc]: done

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- Already signed in https://github.com/cms-sw/cmssw/pull/21146#issuecomment-343552121
- Fine with the implementation of the subsequent update suggested by David

**Comment by kpedro88** [comment]
+1

**Comment by davidlange6** [comment]
merge

**Pull request merged**